### DISCUSSION TITLE
HitObjects added post-beatmap generation (during gameplay)

### DISCUSSION THREAD

**Discussion by LumpBloom7** [ppy/osu#16947] [Ideas]
During gameplay, `HitObjects` can only be sourced from beatmaps which is created ahead of time. Any attempts to create more during gameplay will make the game end before all `HitObjects` are hit/missed as the `ScoreProcessor` doesn't consider newly added `HitObjects`.

This prevents rulesets from incorporating more dynamic gameplay that doesn't have any guarantees on the total number of `HitObject`s that could appear. Such rulesets would be better suited for casual/multiplayer gameplay, rather than to get the max score. 

As an example of a use case: A rechargeable perk that would add a few `HitObject`s to a player's (potentially others) Playfield, increasing hits, that can be activated at any time.

Proposal: `ScoreProcessor` should increment `maxHits` when new `HitObject`s are added to the playfield, so that it doesn't prematurely end the session. The 1 million score baseline could still be calculated using only non-dynamic `HitObject`s.

**Comment by peppy** [comment]
Definitely something which is required, but we'll need to think through how the implementation should work.
